File: rotateArray.cpp
class Solution {
private:
    void swap(int *x, int *y)
    {
        int temp = *x;
        *x = *y;
        *y = temp;
    }

public:
    void rotate(vector<int>& nums, int k) {
        
        int n = nums.size();
        
        if(n == 1)
            return;
        
        k = k % n;
        
        for(int p1 = n - k, p2 = n - 1; p1 < p2; p1++, p2--)
           swap(&nums[p1], &nums[p2]);
        
        
        for(int p1 = 0, p2 = n - k - 1; p1 < p2; p1++, p2--)
             swap(&nums[p1], &nums[p2]);
        
        
        for(int p1 = 0, p2 = n - 1; p1 < p2; p1++, p2--)
             swap(&nums[p1], &nums[p2]);
        
    }
};

File: progress_sheet/rotate_array.py
class Solution:

    def flip_array(self, nums, left, right):
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1

    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        size = len(nums)
        flip_pos = size - (k % size)
        left = 0
        right = flip_pos - 1

        # flip left subarray
        self.flip_array(nums,left,right)
                
        # flip right subarray
        left = flip_pos
        right = size - 1
        
        self.flip_array(nums,left,right)
                
        # Flip the whole array
        nums.reverse()